src/layout/parser.rs

```diff
@@ -242,6 +259,171 @@ where
+                style = FontStyle::Italic;
+            } else if token.eq_ignore_ascii_case("thin") || token.eq_ignore_ascii_case("hairline") {
+                weight = FontWeight::Thin;
+            } else if token.eq_ignore_ascii_case("extralight")
+                || token.eq_ignore_ascii_case("ultralight")
+            {
+                weight = FontWeight::ExtraLight;
+            } else if token.eq_ignore_ascii_case("light") {
+                weight = FontWeight::Light;
+            } else if token.eq_ignore_ascii_case("semilight")
+                || token.eq_ignore_ascii_case("demilight")
+            {
+                weight = FontWeight::SemiLight;
+            } else if token.eq_ignore_ascii_case("normal") {
+                weight = FontWeight::Normal;
+            } else if token.eq_ignore_ascii_case("medium") {
+                weight = FontWeight::Medium;
+            } else if token.eq_ignore_ascii_case("semibold")
+                || token.eq_ignore_ascii_case("demibold")
+            {
+                weight = FontWeight::SemiBold;
+            } else if token.eq_ignore_ascii_case("bold") {
+                weight = FontWeight::Bold;
+            } else if token.eq_ignore_ascii_case("extrabold")
+                || token.eq_ignore_ascii_case("ultrabold")
+            {
+                weight = FontWeight::ExtraBold;
+            } else if token.eq_ignore_ascii_case("black") || token.eq_ignore_ascii_case("heavy") {
+                weight = FontWeight::Black;
+            } else if token.eq_ignore_ascii_case("extrablack")
+                || token.eq_ignore_ascii_case("ultrablack")
+            {
+                weight = FontWeight::ExtraBlack;
+            } else if token.eq_ignore_ascii_case("ultracondensed") {
+                stretch = FontStretch::UltraCondensed;
+            } else if token.eq_ignore_ascii_case("extracondensed") {
+                stretch = FontStretch::ExtraCondensed;
+            } else if token.eq_ignore_ascii_case("condensed") {
+                stretch = FontStretch::Condensed;
+            } else if token.eq_ignore_ascii_case("semicondensed") {
+                stretch = FontStretch::SemiCondensed;
+            } else if token.eq_ignore_ascii_case("normal") {
+                stretch = FontStretch::Normal;
+            } else if token.eq_ignore_ascii_case("semiexpanded") {
+                stretch = FontStretch::SemiExpanded;
+            } else if token.eq_ignore_ascii_case("expanded") {
+                stretch = FontStretch::Expanded;
+            } else if token.eq_ignore_ascii_case("extraexpanded") {
+                stretch = FontStretch::ExtraExpanded;
+            } else if token.eq_ignore_ascii_case("ultraexpanded") {
+                stretch = FontStretch::UltraExpanded;
+            } else if !token.eq_ignore_ascii_case("regular") {
+                family =
+                    &family[..token.as_ptr() as usize - family.as_ptr() as usize + token.len()];
```

Comment:
What does this line do?

Reply:
So this is where we notice that the token we encountered is part of the family name and we should stop scanning. So what we want here is to assign the family name up until and including the token we just looked at. Sometimes iterators in Rust have the ability to give you access to the remaining underlying data, but somehow not .split_whitespace(), but even then it wouldn't include the current item anyway, only the rest. So what I'm doing instead is some pointer shenanigans (still all completely safe). Basically I take the token's pointer and turn it into an integer address and subtract the address of the whole family name. That is how many bytes into the whole string the token is. If we add the length of the token to it (since we want to include it, as it turned out to not be a style token) we get the total length from the beginning of the whole string that we consider the family name. We then take a substring based on that length.